internal/shell/shell_test.go

```diff
@@ -159,12 +162,66 @@ func TestTabCompletionStdlib(t *testing.T) {
 	assertTabCompletionWithPrefix(t, prefix, strlib, "//"+lib+".%s\t", nil)
 }
 
+func TestTrimExpr(t *testing.T) {
+	t.Parallel()
+
+	sh := newShellInstance(newLineCollector(), syntax.StdScope())
+
+	realExpr, residue := sh.trimExpr(`x.`)
+	assert.Equal(t, "x", realExpr)
+	assert.Equal(t, ".", residue)
+
+	realExpr, residue = sh.trimExpr(`abc(`)
+	assert.Equal(t, "abc", realExpr)
+	assert.Equal(t, "(", residue)
+
+	realExpr, residue = sh.trimExpr(`x`)
+	assert.Equal(t, "x", realExpr)
+	assert.Equal(t, "", residue)
+
+	//TODO: more advanced predictions
+	// realExpr, residue = sh.trimExpr(`abc.ab`)
+	// assert.Equal(t, "abc", realExpr)
+	// assert.Equal(t, ".ab", residue)
+}
+
+func TestCompletionCurrentExpr(t *testing.T) {
+	t.Parallel()
+
+	assertTabCompletion(t, []string{".a"}, 0, "(a: 1)\t", nil)
+	assertTabCompletion(t, []string{".a", ".b"}, 0, "(a: 1, b: 2)\t", nil)
+	assertTabCompletion(t, []string{".a", ".b"}, 0, "(a: 1, b: 2)\t + 123", nil)
+	assertTabCompletion(t, []string{"a"}, 1, "(a: 1).\t", nil)
+	assertTabCompletion(t, []string{"a", "b"}, 1, "(a: 1, b: 2).\t", nil)
+	assertTabCompletion(t, []string{"a", "b"}, 1, "(a: 1, b: 2).\t + 123", nil)
+	assertTabCompletion(t, []string{".c"}, 0, "(a: (c: 3), b: 2).a\t", nil)
+	assertTabCompletion(t, []string{`"random string"`}, 1, "(`random string`: 1).\t", nil)
```

Comment:
The string syntax for the suggestions should match what the standard output code does for tuple keys and strings. E.g:

```arrai
@> (a:1, 'b"b': 2, "c'c": 3, "d`d": 4, "e\"e'e`ee": 5)
(a: 1, 'b"b': 2, "c'c": 3, 'd`d': 4, "e\"e'e`ee": 5)
@> ['a', 'b"b', "c'c", "d`d", "e\"e'e`ee"]
['a', 'b"b', "c'c", 'd`d', "e\"e'e`ee"]
```

There's a function somewhere that does this for you.